xdsl/dialects/ml_program.py

```diff
@@ -0,0 +1,158 @@
+from __future__ import annotations
+
+from typing_extensions import Self
+
+from xdsl.dialects.builtin import (
+    StringAttr,
+    SymbolRefAttr,
+    UnitAttr,
+)
+from xdsl.ir import (
+    Attribute,
+    Dialect,
+    TypeAttribute,
+)
+from xdsl.irdl import (
+    IRDLOperation,
+    attr_def,
+    irdl_op_definition,
+    result_def,
+)
+from xdsl.parser import Parser
+from xdsl.printer import Printer
+from xdsl.utils.exceptions import VerifyException
+
+
+@irdl_op_definition
+class Global(IRDLOperation):
+    """
+    Module level declaration of a global variable
+
+    See https://mlir.llvm.org/docs/Dialects/MLProgramOps/#ml_programglobal-ml_programglobalop
+    """
+
+    name = "ml_program.global"
+
+    sym_name = attr_def(StringAttr)
+    type = attr_def(TypeAttribute)
+    is_mutable = attr_def(UnitAttr)
+    value = attr_def(Attribute)
+    sym_visibility = attr_def(StringAttr)
+
+    def __init__(
+        self,
+        sym_name: Attribute,
+        type: Attribute,
+        is_mutable: Attribute,
+        value: Attribute,
+        sym_visibility: Attribute,
+    ):
+        super().__init__(
+            attributes={
+                "sym_name": sym_name,
+                "type": type,
+                "is_mutable": is_mutable,
+                "value": value,
+                "sym_visibility": sym_visibility,
+            },
+        )
+
+    def print(self, printer: Printer):
+        printer.print_string(" ")
+        if self.sym_visibility:
+            printer.print_string(self.sym_visibility.data)
+        printer.print_string(" ")
+        printer.print_string("@")
+        printer.print_string(self.sym_name.data)
+        printer.print_string("(")
+        printer.print_attribute(self.value)
+        printer.print_string(")")
+        printer.print_string(" : ")
+        printer.print_attribute(self.type)
+
+    @classmethod
+    def parse(cls, parser: Parser) -> Self:
+        attrs = parser.parse_optional_attr_dict()
+        if parser.parse_optional_keyword("public"):
+            sym_visibility = StringAttr("public")
+        elif parser.parse_optional_keyword("nested"):
+            sym_visibility = StringAttr("nested")
+        elif parser.parse_optional_keyword("private"):
+            sym_visibility = StringAttr("private")
+        else:
+            sym_visibility = StringAttr("")
+        is_mutable = UnitAttr()
+        sym_name = parser.parse_symbol_name().data
+        if parser.parse_optional_punctuation("("):
```

Comment:
I don't think this is optional

Reply:
why isnt it, we can have a case like this: ```ml_program.global mutable @foobar : tensor<?xi32>```